File: cpp/cards/golf/game_state.cc
#include "cpp/cards/golf/game_state.h"

#include <algorithm>
#include <unordered_set>
#include <utility>
#include <vector>

#include "absl/status/statusor.h"
#include "cpp/cards/card.h"
#include "cpp/cards/golf/player.h"

namespace golf {
using namespace cards;
using absl::FailedPreconditionError;
using absl::StatusOr;

using std::deque;
using std::string;
using std::unordered_map;
using std::unordered_set;
using std::vector;

bool GameState::isOver() const { return drawPile.empty() || whoseTurn == whoKnocked; }

bool GameState::allPlayersPresent() const {
  return std::all_of(players.begin(), players.end(), [](const Player& p) { return p.isPresent(); });
}

unordered_set<int> GameState::winners() const {
  unordered_set<int> winningPlayers;
  int minScore = 40;  // max score is 9 10 Q K == 39
  int playerIndex = 0;
  for (auto& p : players) {
    int playerScore = p.score();
    if (playerScore < minScore) {
      minScore = playerScore;
      winningPlayers.clear();
    }
    if (playerScore == minScore) {
      winningPlayers.insert(playerIndex);
    }

    playerIndex++;
  }
  if (winningPlayers.find(whoKnocked) != winningPlayers.end()) {
    winningPlayers.clear();
    winningPlayers.insert(whoKnocked);
  }

  return winningPlayers;
}

StatusOr<GameState> GameState::peekAtDrawPile(int player) const {
  if (isOver()) {
    return FailedPreconditionError("game is over");
  }
  if (!allPlayersPresent()) {
    return FailedPreconditionError("not all players have joined");
  }
  if (whoseTurn != player) {
    return FailedPreconditionError("not your turn");
  }
  if (peekedAtDrawPile) {
    return FailedPreconditionError("you can only peek once per turn");
  }

  return GameState{drawPile, discardPile, players, true, whoseTurn, whoKnocked, gameId};
}

StatusOr<GameState> GameState::swapDrawForDiscardPile(int player) const {
  if (isOver()) {
    return FailedPreconditionError("game is over");
  }
  if (!allPlayersPresent()) {
    return FailedPreconditionError("not all players have joined");
  }
  if (whoseTurn != player) {
    return FailedPreconditionError("not your turn");
  }

  // update draw pile
  deque<Card> updatedDrawPile{drawPile};
  Card toSwampIntoDiscard = updatedDrawPile.back();
  updatedDrawPile.pop_back();
  const deque<Card> drawPileForNewGameState = std::move(updatedDrawPile);

  deque<Card> updatedDiscardPile{discardPile};
  updatedDiscardPile.push_back(toSwampIntoDiscard);
  const deque<Card> discardPileForNewGameState = std::move(updatedDiscardPile);

  // update whose turn it is
  int newWhoseTurn = (whoseTurn + 1) % players.size();

  return GameState{drawPileForNewGameState,
                   discardPileForNewGameState,
                   players,
                   false,
                   newWhoseTurn,
                   whoKnocked,
                   gameId};
}

StatusOr<GameState> GameState::swapForDrawPile(int player, Position position) const {
  if (isOver()) {
    return FailedPreconditionError("game is over");
  }
  if (!allPlayersPresent()) {
    return FailedPreconditionError("not all players have joined");
  }
  if (whoseTurn != player) {
    return FailedPreconditionError("not your turn");
  }

  // update draw pile
  deque<Card> updatedDrawPile{drawPile};
  Card toSwampIntoHand = updatedDrawPile.back();
  updatedDrawPile.pop_back();
  const deque<Card> drawPileForNewGameState = std::move(updatedDrawPile);

  // update current player
  const Player currentPlayer = players.at(player);
  Card toSwapOutOfHand = currentPlayer.cardAt(position);
  const Player updatedCurrentPlayer = currentPlayer.swapCard(toSwampIntoHand, position);

  // update players list
  vector<Player> updatedPlayers;
  for (size_t i = 0; i < players.size(); i++) {
    if (static_cast<int>(i) == whoseTurn) {
      updatedPlayers.push_back(updatedCurrentPlayer);
    } else {
      updatedPlayers.push_back(players.at(i));
    }
  }
  const vector<Player> playersForNewGameState = std::move(updatedPlayers);

  // update discard pile
  deque<Card> updatedDiscardPile{discardPile};
  updatedDiscardPile.push_back(toSwapOutOfHand);
  const deque<Card> discardPileForNewGameState = std::move(updatedDiscardPile);

  // update whose turn it is
  int newWhoseTurn = (whoseTurn + 1) % players.size();

  return GameState{drawPileForNewGameState,
                   discardPileForNewGameState,
                   playersForNewGameState,
                   false,
                   newWhoseTurn,
                   whoKnocked,
                   gameId};
}

absl::StatusOr<GameState> GameState::swapForDiscardPile(int player, Position position) const {
  if (isOver()) {
    return FailedPreconditionError("game is over");
  }
  if (!allPlayersPresent()) {
    return FailedPreconditionError("not all players have joined");
  }
  if (whoseTurn != player) {
    return FailedPreconditionError("not your turn");
  }
  if (peekedAtDrawPile) {
    return FailedPreconditionError("cannot swap for discard after peeking");
  }

  // remove top card from discard pile
  deque<Card> mutableDiscardPile{discardPile};

  // TODO: how should we enforce looking at the card once?
  Card toSwampIntoHand = mutableDiscardPile.back();
  mutableDiscardPile.pop_back();

  // update current player
  const Player currentPlayer = players.at(player);
  Card toSwapOutOfHand = currentPlayer.cardAt(position);
  const Player updatedCurrentPlayer = currentPlayer.swapCard(toSwampIntoHand, position);

  // update discardPile
  mutableDiscardPile.push_back(toSwapOutOfHand);
  const deque<Card> discardPileForNewGameState = std::move(mutableDiscardPile);

  // update players list
  vector<Player> updatedPlayers;
  for (size_t i = 0; i < players.size(); i++) {
    if (static_cast<int>(i) == whoseTurn) {
      updatedPlayers.push_back(updatedCurrentPlayer);
    } else {
      updatedPlayers.push_back(players.at(i));
    }
  }
  const vector<Player> playersForNewGameState = std::move(updatedPlayers);

  // update whose turn it is
  int newWhoseTurn = (whoseTurn + 1) % players.size();

  return GameState{
      drawPile, discardPileForNewGameState, playersForNewGameState, false, newWhoseTurn, whoKnocked,
      gameId};
}

StatusOr<GameState> GameState::knock(int player) const {
  if (isOver()) {
    return FailedPreconditionError("game is over");
  }
  if (!allPlayersPresent()) {
    return FailedPreconditionError("not all players have joined");
  }
  if (whoseTurn != player) {
    return FailedPreconditionError("not your turn");
  }
  if (peekedAtDrawPile) {
    return FailedPreconditionError("cannot knock after peeking");
  }
  if (whoKnocked != -1) {
    return FailedPreconditionError("someone already knocked");
  }

  // update whose turn it is
  int newWhoseTurn = (whoseTurn + 1) % players.size();

  return GameState{drawPile, discardPile, players, false, newWhoseTurn, player, gameId};
}

GameState GameState::withPlayers(vector<Player> newPlayers) const {
  return GameState{drawPile,   discardPile, std::move(newPlayers), false, whoseTurn,
                   whoKnocked, gameId};
}

}  // namespace golf

File: go/cards/golf/game_state.go
package golf

import (
	"errors"
	"github.com/muchq/moonbase/go/cards"
	"slices"
)

type GameState struct {
	Id               string
	Version          string
	Players          []Player
	PeekedAtDrawPile bool
	WhoseTurn        int
	WhoKnocked       int
	DrawPile         cards.Deck
	DiscardPile      cards.Deck
}

func (g *GameState) IsOver() bool {
	return g.DrawPile.IsEmpty() || g.WhoseTurn == g.WhoKnocked
}

func (g *GameState) AllPlayersPresent() bool {
	for _, player := range g.Players {
		if player.Id == "" {
			return false
		}
	}
	return true
}

func (g *GameState) Winners() []int {
	winningPlayerIndexes := make([]int, 0)
	minScore := 40 // max score is 9 10 Q K == 39
	for index, player := range g.Players {
		pScore := player.Score()
		if pScore < minScore {
			minScore = pScore
			winningPlayerIndexes = nil
		}
		if pScore == minScore {
			winningPlayerIndexes = append(winningPlayerIndexes, index)
		}
	}

	// Tie goes to the runner
	if slices.Contains(winningPlayerIndexes, g.WhoKnocked) {
		winningPlayerIndexes = nil
		winningPlayerIndexes = append(winningPlayerIndexes, g.WhoKnocked)
	}

	return winningPlayerIndexes
}

func (g *GameState) PeekAtDrawPile(player int) (*GameState, error) {
	if g.IsOver() {
		return nil, errors.New("game is over")
	}

	if !g.AllPlayersPresent() {
		return nil, errors.New("not all players have joined")
	}

	if g.WhoseTurn != player {
		return nil, errors.New("not your turn")
	}

	if g.PeekedAtDrawPile {
		return nil, errors.New("you can only peek once per turn")
	}

	var gameState = *g
	gameState.PeekedAtDrawPile = true

	return &gameState, nil
}

func (g *GameState) SwapDrawForDiscard(player int) (*GameState, error) {
	if g.IsOver() {
		return nil, errors.New("game is over")
	}
	if !g.AllPlayersPresent() {
		return nil, errors.New("not all players have joined")
	}
	if g.WhoseTurn != player {
		return nil, errors.New("not your turn")
	}

	// update draw pile
	var updatedDrawPile = g.DrawPile
	updatedDrawPtr := &updatedDrawPile
	toSwapIntoDiscard := updatedDrawPtr.RemoveLast()

	// update discard pile
	var updatedDiscardPile = g.DiscardPile
	updatedDiscardPtr := &updatedDiscardPile
	updatedDiscardPtr.AddLast(toSwapIntoDiscard)

	// update whose turn it is
	newWhoseTurn := (g.WhoseTurn + 1) % len(g.Players)

	return &GameState{
		DrawPile:         updatedDrawPile,
		DiscardPile:      updatedDiscardPile,
		Id:               g.Id,
		Version:          g.Version,
		Players:          g.Players,
		PeekedAtDrawPile: false,
		WhoseTurn:        newWhoseTurn,
		WhoKnocked:       -1,
	}, nil
}

func (g *GameState) SwapForDrawPile(player int, position Position) (*GameState, error) {
	if g.IsOver() {
		return nil, errors.New("game is over")
	}
	if !g.AllPlayersPresent() {
		return nil, errors.New("not all players have joined")
	}
	if g.WhoseTurn != player {
		return nil, errors.New("not your turn")
	}

	// update draw pile
	var updatedDrawPile = g.DrawPile
	updatedDrawPtr := &updatedDrawPile
	toSwapIntoHand := updatedDrawPtr.RemoveLast()

	// update current player
	var currentPlayer = g.Players[player]
	currentPlayerPtr := &currentPlayer
	toSwapOutOfHand := currentPlayerPtr.CardAt(position)
	currentPlayerPtr.SwapIntoHand(toSwapIntoHand, position)
	g.Players[player] = *currentPlayerPtr

	// update discard pile
	var updatedDiscardPile = g.DiscardPile
	updatedDiscardPtr := &updatedDiscardPile
	updatedDiscardPtr.AddLast(toSwapOutOfHand)

	// update whose turn it is
	newWhoseTurn := (g.WhoseTurn + 1) % len(g.Players)

	return &GameState{
		DrawPile:         updatedDrawPile,
		DiscardPile:      updatedDiscardPile,
		Id:               g.Id,
		Version:          g.Version,
		Players:          g.Players,
		PeekedAtDrawPile: false,
		WhoseTurn:        newWhoseTurn,
		WhoKnocked:       -1,
	}, nil
}
